usbutils 007 (new formula): provides 'lsusb' tool

mikhailai opened the pull request:
The 'lsusb' tools allows listing connected USB devices and
printing detailed information about the devices' USB descriptors,
such as configurations, interfaces, and endpoints.
These details are not available from standard
Mac OS tools, such as ioreg or system_profiler.

Note, the 'usbutils' is available on Macports, and it would be great to have it in the Homebrew as well.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
 -- YES
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
 -- YES
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
 -- YES
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
  -- YES
-----

mikhailai commented:
Jenkins reports a failure, but it seems bogus. Can it be because the commit has a bottle for Sierra that I've built locally?

mistydemeo commented:
> Can it be because the commit has a bottle for Sierra that I've built locally?

We don't accept outside bottles; we'll build our own and include them. Can you drop the bottle block please?

mikhailai commented:
Sure, updated the file.
I was actually confused by the fact that "brew audit --strict --online usbutils" would complain about the missing "bottle" block, so I added it. Looks like I'm only supposed to run "brew audit --new-formula usbutils" on the new formulas.

mistydemeo commented:
Hm, looks like the test failed on 10.11.

```
17:48:56 ==> /usr/local/Cellar/usbutils/007/bin/lsusb -v
17:48:56 Error: usbutils: failed
```

mikhailai commented:
Yep, and I have no clue why. I did not see any messages coming from "lsusb" itself, and it works fine on my El Capitan system. Maybe there is a problem accessing USB on the machines Jenkins is running (because of VM? Some kind of sandboxing/permissions issues?)
I can replace the current tests with a "bad" one (lsusb --version). Do you have any recommendations?

mikhailai commented:
FYI: previously, the same test code ran fine on El Capitan:
https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/27681/
So, it could really be machine-dependent, which probably means we cannot run these tests anyway.

mikhailai commented:
Ok, I double-checked it on AWS Linux, and 'lsusb' seems to not be VM-friendly: it exits with status of 1. I changed the tests to be less strict. Please take another look.

mikhailai commented:
Ping. All checks pass, it's ready to proceed.

mikhailai commented:
Please, please, someone take a look at this. All things should be good now, AFAIK.

This tool really helps to work with USB devices (for example, debug issues in https://joshuawise.com/horndis), and I, being a Homebrew user, don't really want to ask people to install Macports in order to get the "lsusb" and send me their USB information.

laris commented:
+1 support to add.

mikhailai commented:
@laris, if you need the lsusb, you can get it from my tap:
brew install mikhailai/misc/usbutils
I don't think it'll get into homebrew core any time soon.

DomT4 commented on Formula/usbutils.rb:
```diff
@@ -0,0 +1,43 @@
+class Usbutils < Formula
+  # Some of the information is based on Macports 'usbutils':
```
Can we condense the level of code comments here a bit, please?

IMO the only necessary comments are the lines about why it is pinned on the version it is.

mikhailai replied:
Done: cleaned up the comments.

DomT4 commented:
> I don't think it'll get into homebrew core any time soon.

We're a fairly small team of volunteers with upwards of 100+ PRs/Issues to work on at any given moment. Sometimes these things take a little time. Doing our best to get through everything whilst still eating/sleeping/existing, heh.

fxcoudert commented:
```
  # The latest version we can use is 007 (from 2013), because the later
   # versions depend on 'libudev': part of Systemd and not in Homebrew.
   url "https://mirrors.edge.kernel.org/pub/linux/utils/usb/usbutils/usbutils-007.tar.xz"
```

This goes against the criteria for Homebrew core inclusion: https://docs.brew.sh/Acceptable-Formulae because this is not actively maintained software (well, it is, but we can't use newer versions).

fxcoudert commented:
Proposing to ship an alternative implementation to `lsusb`, based on native tools.

mikhailai commented:
That implementation is mostly useless, because it does not provide information about USB interfaces and endpoints.

The real use of "lsusb" is to run "lsusb -v" to get detailed information about a USB device. jlhonora's implementation, in that case, just runs "system_profiler SPUSBDataType" and verbally prints its output: about useful as a bash alias.

Currently, there are NO MacOS native tools, as far as I'm aware, that can give you as much info as 'lsusb'. There used to be USBProber from Apple, but the it was last released for MacOS 10.9, if I'm not mistaken.

DomT4 commented:
> This goes against the criteria for Homebrew core inclusion: https://docs.brew.sh/Acceptable-Formulae because this is not actively maintained software (well, it is, but we can't use newer versions).

Aye. I'm sort of open to looking the other way here given it is already in MacPorts & seems to have a very low maintenance burden, but if people end up leaning towards rejecting this I'm okay with that.

mikhailai commented:
> This goes against the criteria for Homebrew core inclusion: https://docs.brew.sh/Acceptable-Formulae because this is not actively maintained software (well, it is, but we can't use newer versions).

True, but I'd still argue for inclusion, because:
 - The tools is very useful as-is
 - I don't think there are any significant issues with version 007 resolved in subsequent versions.
 - Security-wise, the tool is safe: it is normally executed as user's process, it does not perform network communication, etc.

mikhailai commented:
@DomT4 , I fixed the comments. Please let me know if the formula needs anything else. For your reference:
https://stackoverflow.com/questions/17058134/is-there-an-equivalent-of-lsusb-for-os-x

Some people are fine with jlhonora's implementation, others are not. My answer (this formula installed from my tap) has already gotten 5 upvotes, being there for less than 2 weeks on a 5-years-old question.

fxcoudert commented on Formula/usbutils.rb:
```diff
@@ -0,0 +1,37 @@
+                          "--prefix=#{prefix}"
+    # Clear/override several variables, to avoid building/installing
+    # linux-specific utilities that rely on sysfs.
+    system "make", "install", "SUBDIRS=", "AM_LDFLAGS=", "bin_SCRIPTS=",
```
Why is `AM_LDFLAGS` there? I don't think it has to do with sysfs

mikhailai replied:
Inaccurate comment. Updated. Please take another look.
AM_LDFLAGS is needed because:

usbutils-007/Makefile.am, line 4:

AM_LDFLAGS = \
	-Wl,--as-needed

The MacOS linker does not support "--as-needed" flag, so we need to override it from the command line to avoid the linker errors. Macports does the same, BTW.

fxcoudert commented:
@DomT4 I won't object too hard. I find it weird that, if people really need a linux-like `lsusb`, they use outdated/unmaintained software rather than write a libusb-based solution (possibly based on the older `usbutils` codebase). But, well, that's life ;)

mikhailai commented:
Fun fact: CentOS 7, Debian Stretch, and Ubuntu Bionic (and not yet released Cosmic) are all using "usbutils" - version 007. Apparently, no one bothers with further versions.

@fxcoudert, well people have already written a libusb-based solution ages ago, and that's what I'm trying to include here ;).

One could fork the "usbutils" codebase at version 007 and maintain it for Mac, but there has probably been no need to do it at that moment.

DomT4 commented:
Given my lack of firm opinion and @fxcoudert's firmer opinion it's likely we'll go with @fxcoudert unless anyone else weighs in. I'll give it a day or so to see if a majority forms in the other direction but if not our advice will likely be to consider supporting this in a tap yourself, which is a pretty easy thing to do.

DomT4 commented:
There's not been any stated interest from other maintainers on swinging the pendulum the other way on this so I'm going to go with the only maintainer who has expressed a firm opinion here and that's @fxcoudert. Misty is more than welcome to correct me if her earlier comments were indicative of more support than I'm realising.

Barring that, I think this is something we're going to have to reject packaging in core at this point in time, which I appreciate isn't the decision you were hoping for. As I mentioned previously there's nothing stopping you maintaining this in a [tap yourself](https://github.com/Homebrew/brew/blob/master/docs/How-to-Create-and-Maintain-a-Tap.md) which is a pretty easy process. Thank you nonetheless for the submission @mikhailai.
